src/dbnode/storage/index.go

```diff
@@ -1554,52 +1605,73 @@ func (i *nsIndex) queryWithSpan(
-		// the loop.
-		seriesCount := results.Size()
-		docsCount := results.TotalDocsCount()
-		alreadyExceededLimit := opts.SeriesLimitExceeded(seriesCount) || opts.DocsLimitExceeded(docsCount)
-		if alreadyExceededLimit {
-			state.Lock()
-			state.exhaustive = false
-			state.Unlock()
-			// Break out if already not exhaustive.
-			break
+	blockIters := make([]*blockIter, len(blocks))
+	for b, block := range blocks {
+		iter, err := newBlockIterFn(ctx, block, query, results)
+		if err != nil {
+			return false, err
+		}
+		blockIters[b] = &blockIter{
+			iter:       iter,
+			iterCloser: x.NewSafeCloser(iter),
+			block:      block,
+		}
+	}
+
+	defer func() {
+		for _, iter := range blockIters {
+			// safe to call Close multiple times, so it's fine to eagerly close in the loop below and here.
+			_ = iter.iterCloser.Close()
 		}
+	}()
+
+	// used by each blockIter to break the iteration loop if the query can't continue processing.
+	breakIter := func() bool {
+		return opts.LimitsExceeded(results.Size(), results.TotalDocsCount()) || state.hasErr()
+	}
 
-		// Calculate time spent waiting for a worker
+	// We're looping through all the blocks that we need to query and kicking
+	// off parallel queries which are bounded by the permits maximum
+	// concurrency. It's possible at this point that we've completed querying one or more blocks and already exhausted
+	// the maximum number of results that we're allowed to return. If thats the case, there is no value in kicking off
+	// more parallel queries, so we break out of the loop.
+	for _, blockIter := range blockIters {
+		// Capture for async query execution below.
+		blockIter := blockIter
 		wg.Add(1)
-		scheduleResult := i.queryWorkersPool.GoWithContext(ctx, func() {
-			startProcessing := time.Now()
-			execBlockFn(ctx, block, query, opts, &state, results, logFields)
-			i.metrics.queryMetrics.blockProcessingTime.RecordDuration(time.Since(startProcessing))
-			wg.Done()
-		})
-		totalWaitTime += scheduleResult.WaitTime
-		if !scheduleResult.Available {
-			state.Lock()
-			state.multiErr = state.multiErr.Add(index.ErrCancelledQuery)
-			state.Unlock()
-			// Did not launch task, need to ensure don't wait for it
+		go func() {
+			for !blockIter.iter.Done() && !breakIter() {
+				startWait := time.Now()
+				err := permits.Acquire(ctx)
+				blockIter.waitTime += time.Since(startWait)
+				if err != nil {
+					state.addErr(err)
+					break
+				}
+				// check the query hasn't been canceled while waiting.
+				if !breakIter() {
+					startProcessing := time.Now()
+					execBlockFn(ctx, blockIter.block, blockIter.iter, opts, &state, results, logFields)
+					processingTime := time.Since(startProcessing)
+					i.metrics.queryMetrics.blockProcessingTime.RecordDuration(processingTime)
+					blockIter.processingTime += processingTime
+				}
+				permits.Release()
+			}
+			blockIter.searchTime += blockIter.iter.SearchDuration()
```

Comment:
a nice plus of this change is we no longer have to acquire locks to update the timing info on the shared results. each goroutine has its own timing information and the sub results are added together when all queries are done.

Reply:
+1